pkg/controller/snapshotter/snapshotter.go

```diff
@@ -755,6 +865,48 @@ func (vs *volumeSnapshotter) UpdateVolumeSnapshot(snapshotName string, status *[
 	return &cloudTags, nil
 }
 
+func (vs *volumeSnapshotter) bindVolumeSnapshotDataToVolumeSnapshot(snapshotName string, snapshotDataName string) error {
+        var snapshotObj crdv1.VolumeSnapshot
+
+        glog.Infof("In bindVolumeSnapshotDataToVolumeSnapshot")
+        // Get a fresh copy of the VolumeSnapshot from the API server
+        snapNameSpace, snapName, err := cache.GetNameAndNameSpaceFromSnapshotName(snapshotName)
+        if err != nil {
+                return fmt.Errorf("Error getting namespace and name from VolumeSnapshot name %s: %v", snapshotName, err)
+        }
+        glog.Infof("bindVolumeSnapshotDataToVolumeSnapshot: Namespace %s Name %s", snapNameSpace, snapName)
+        err = vs.restClient.Get().
+                Name(snapName).
+                Resource(crdv1.VolumeSnapshotResourcePlural).
+                Namespace(snapNameSpace).
+                Do().Into(&snapshotObj)
+
+        objCopy, err := vs.scheme.DeepCopy(&snapshotObj)
+        if err != nil {
+                return fmt.Errorf("Error copying snapshot object %s object from API server: %v", snapshotName, err)
+        }
+        snapshotCopy, ok := objCopy.(*crdv1.VolumeSnapshot)
+        if !ok {
+                return fmt.Errorf("Error: expecting type VolumeSnapshot but received type %T", objCopy)
+        }
+
+	snapshotCopy.Spec.SnapshotDataName = snapshotDataName
+        glog.Infof("bindVolumeSnapshotDataToVolumeSnapshot: Updating VolumeSnapshot object [%#v]", snapshotCopy)
+        // TODO: Make diff of the two objects and then use restClient.Patch to update it
+        var result crdv1.VolumeSnapshot
+        err = vs.restClient.Put().
```

Comment:
I think this will trigger update event on the VolumeSnapshot and since we're adding VolumeSnapshotData we may end up with duplicate snapshots: https://github.com/rootfs/snapshot/blob/ddb206bd9cf755aba4b8c1b0fe42cc67d66a50fd/pkg/controller/snapshot-controller/snapshot-controller.go#L202

Reply:
I see the following on line 207:
	if oldSnapshot.Spec.SnapshotDataName != newSnapshot.Spec.SnapshotDataName {
		c.desiredStateOfWorld.AddSnapshot(newSnapshot)
	}
So as long as the SnapshotDataName is not changed, we should not have duplicate snapshots?

Reply:
Yes. But when we're adding a new VolumeSnapshotData, the name changes (from "" to something else). I think the problem is in the VolumeSnapshot update event handler. I'll try to run some more tests.

Reply:
Silly me... The controller will try to add a new snapshot. But as long as the VolumeSnapshot name doesn't change, no duplicates can appear. So it's harmless.